### PR TITLE
Using 'after-emit' hook instead of 'done' so that other plugins can read the files that this plugin has written to the filesystem

### PR DESCRIPTION
**Pull request by max-b**
Let me know if this isn't of interest to you. It's possible that I have a unique use-case, but I wanted this plugin to write the assets to disk before another plugin had it's `after-emit` event triggered. Specifically, using `memory-fs` along with `workbox-webpack-plugin`. `workbox-webpack-plugin` depends on the files existing on the disk in order to generate appropriate service worker files.

Also: I had to change the 'build' npm script to 'prepare' so that it can be npm installed from a git repo instead of having to be published as an npm repo.

And had to adjust eslint rules a tiny bit.

Thanks!

### CODE REVIEW

**Comment by ghost** [comment]
@max-b, thank you so much! Your use-case is certainly not unique, I too need some other plugins to be able to read emitted files. Will be using a fork of your PR while it's not merged. Thanks again, it really helped!

**Comment by max-b** [comment]
@st-sloth glad to hear it's helpful :grinning: I wasn't sure if anyone would run into the same use case. @gajus any interest in merging this? It looks like it's maybe useful to other folks and I don't think it conflicts anywhere else?

File: .eslintrc [hunk 1]
```diff
@@ -1,3 +1,6 @@
 {
-  "extends": "canonical"
+  "extends": "canonical",
+  "rules": {
+    "promise/prefer-await-to-callbacks": "off",
```
Comment (gajus) [inline .eslintrc]: No need to disable the rule across the project. Use `// eslint-disable-next-line promise/prefer-await-to-callbacks`

File: package.json [hunk 2]
```diff
@@ -25,7 +25,7 @@
   },
   "scripts": {
     "lint": "eslint ./src",
-    "build": "babel --source-maps --copy-files ./src --out-dir ./dist",
+    "prepare": "babel --source-maps --copy-files ./src --out-dir ./dist",
```
Comment (gajus) [inline package.json]: Keep "build".

File: src/index.js [hunk 3]
```diff
@@ -110,18 +110,22 @@ export default (userOptions: UserOptionsType = {}): Object => {
       return setupStatus;
     };
 
-    compiler.plugin('done', (stats) => {
+    compiler.plugin('after-emit', (compilation, callback) => {
       if (!setup()) {
+        callback('write-file-webpack-plugin couldn\'t setup.');
```
Comment (gajus) [inline src/index.js]: Use error as a callback, i.e. `callback(new Error('write-file-webpack-plugin couldn\'t setup'););`

**Review by gajus — CHANGES_REQUESTED** [review]

File: src/index.js (hunk 3 above)
Comment (max-b) [reply src/index.js]: Good call :smiley:

File: package.json (hunk 2 above)
Comment (max-b) [reply package.json]: gotcha

File: .eslintrc (hunk 1 above)
Comment (max-b) [reply .eslintrc]: Ok - there's a few of them, but I don't think it ends up being too cluttered.

**Comment by max-b** [comment]
I had to merge the new webpack 4 style plugin code in. I think I got it right, though I will admit I haven't tested it against webpack 4. I'm not super familiar with best practices for testing webpack plugins though?

**Review by gajus — APPROVED** [review]

**Comment by gajus** [comment]
Looks good to me.

**Pull request merged**